Python/ceval.c

```diff
@@ -4603,10 +4639,42 @@ _PyEval_EvalFrameDefault(PyThreadState *tstate, InterpreterFrame *frame, int thr
 
         TARGET(CALL_FUNCTION): {
             PREDICTED(CALL_FUNCTION);
-            PyObject **sp, *res;
-            sp = stack_pointer;
-            res = call_function(tstate, &sp, oparg, NULL, cframe.use_tracing);
-            stack_pointer = sp;
+            PyObject *res;
+
+            // Check if the call can be inlined or not
+            PyObject *function = PEEK(oparg + 1);
+            if (Py_TYPE(function) == &PyFunction_Type) {
+                PyCodeObject *code = (PyCodeObject*)PyFunction_GET_CODE(function);
+                STACK_SHRINK(oparg + 1);
+                if ((code->co_flags & (CO_GENERATOR | CO_COROUTINE | CO_ASYNC_GENERATOR)) == 0) {
+                    InterpreterFrame *new_frame = _PyEval_FrameFromPyFunctionAndArgs(tstate, stack_pointer+1, oparg, function);
+                    if (new_frame == NULL) {
+                        // When we exit here, we own all variables in the stack (the frame creation has not stolen
```

Comment:
Wait, since you called ``STACK_SHRINK(oparg+1)`` above, don't you need to ``Py_DECREF`` all the function arguments and function objects too? I don't think ``error`` will decref them automatically for you.

Reply:
That's true (and my fault).

We need to make functions that steal references, always steal those references.
Otherwise this sort of error is just too easy to make.